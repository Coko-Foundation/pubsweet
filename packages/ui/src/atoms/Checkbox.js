import React from 'react'
import styled, { keyframes } from 'styled-components'
import th from '../helpers/themeHelper'

const checking = keyframes`
  0% {
    transform: scale(0.8);
  }
  20% {
    transform: scale(1.2);
  }
  80% {
    transform: scale(1);
  }
  100% {
    transform: scale(1);
  }
`

const localBorderSize = '3px'
const localBorderTwoSize = '1px'

const Root = styled.label`
  align-items: center;
  display: ${props => (props.inline ? 'inline-flex' : 'flex')};
  font-family: ${th('fontAuthor')};
  font-style: italic;
  letter-spacing: 1px;
  transition: all 2s;

  &:not(:last-child) {
    margin-right: ${props => (props.inline ? '2.7em' : '0')};
  }

  & input {
    display: none;
    margin-right: 0.25rem;
  }

  & span {
    font-size: 1.1em;
    transition: color 0.5s;
  }

  &:hover span {
    color: ${th('colorPrimary')};
  }

  & span::before {
    content: ' ';
    display: inline-block;
    vertical-align: middle;

    width: 0.7em;
    height: 0.7em;
    margin-right: 0.5em;

    background: ${props => (props.checked ? 'currentcolor' : 'transparent')};
    border: ${localBorderSize} solid white;
    box-shadow: 0 0 0 ${localBorderTwoSize} currentcolor;

    transition: border 0.5s ease, background-size 0.3s ease;
  }

  &:hover span::before {
    animation: ${checking} 0.5s;

    background: ${th('colorPrimary')};
    box-shadow: 0 0 0 ${localBorderTwoSize} ${th('colorPrimary')};
  }
`

const Checkbox = ({
  inline,
  name,
  value,
  label,
  checked,
  required,
  onChange,
}) => {
  checked = checked || false

  return (
    <Root checked={checked} inline={inline}>
      <input
        checked={checked}
        name={name}
        onChange={onChange}
        required={required}
        type="checkbox"
        value={value}
      />
      <span>{label}</span>
    </Root>
  )
}

export default Checkbox
