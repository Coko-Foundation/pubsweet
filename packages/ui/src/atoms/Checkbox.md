A checkbox.

```js
import { useState } from 'react'
const [state, setState] = useState({ checked: null })

;<Checkbox
  name="checkbox"
  checked={state.checked}
  onChange={event => setState({ checked: event.target.checked })}
/>
```

A checked checkbox.

```js
import { useState } from 'react'
const [state, setState] = useState({ checked: true })

;<Checkbox
  name="checkbox-checked"
  checked={state.checked}
  onChange={event => setState({ checked: event.target.checked })}
/>
```

A checkbox with a label.

```js
import { useState } from 'react'
const [state, setState] = useState({ checked: false })

;<Checkbox
  name="checkbox-labelled"
  checked={state.checked}
  label="Foo"
  onChange={event => setState({ checked: event.target.checked })}
/>
```
