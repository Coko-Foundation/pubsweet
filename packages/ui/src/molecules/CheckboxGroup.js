import React from 'react'
import { clone } from 'lodash'

import Checkbox from '../atoms/Checkbox'
import Flexbox from '../atoms/Flexbox'

class CheckboxGroup extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      values: props.value ? clone(props.value) : [],
    }
  }

  handleChange = event => {
    const values = [...this.state.values]

    const { value } = event.target

    if (event.target.checked) {
      values.push(value)
    } else {
      values.splice(values.indexOf(value), 1)
    }

    this.setState({ values })

    this.props.onChange(values)
  }

  render() {
    const { inline, name, options, required } = this.props
    const { values } = this.state

    return (
      <Flexbox column={!inline}>
        {options.map(option => (
          <Checkbox
            checked={values.includes(option.value)}
            inline={inline}
            key={option.value}
            label={option.label}
            name={name}
            onChange={this.handleChange}
            required={required}
            value={option.value}
          />
        ))}
      </Flexbox>
    )
  }
}

export default CheckboxGroup
