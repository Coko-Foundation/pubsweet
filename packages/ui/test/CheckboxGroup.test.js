import React from 'react'
import { clone } from 'lodash'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import 'jest-styled-components'

import Checkbox from '../src/atoms/Checkbox'
import CheckboxGroup from '../src/molecules/CheckboxGroup'

const props = {
  options: [
    {
      label: 'One',
      value: 'one',
    },
    {
      label: 'Two',
      value: 'two',
    },
    {
      label: 'Three',
      value: 'three',
    },
  ],
}

const wrapper = shallow(<CheckboxGroup {...props} />)
const Checkboxes = wrapper.find(Checkbox)

describe('Checkbox Group', () => {
  test('is rendered correctly', () => {
    const tree = renderer.create(<CheckboxGroup {...props} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('Renders a number of Checkbox buttons', () => {
    const len = props.options.length
    expect(wrapper.children()).toHaveLength(len)

    let i = 0

    while (i < len) {
      const child = wrapper.childAt(i)
      expect(child.is(Checkbox)).toBeTruthy()

      i += 1
    }

    expect(Checkboxes).toHaveLength(len)
  })

  test('Checkboxes get the correct props', () => {
    const CheckboxComps = Checkboxes.getElements()
    let i = 0

    while (i < props.options.length) {
      const Checkbox = CheckboxComps[i]
      const CheckboxProps = Checkbox.props

      expect(CheckboxProps.label).toEqual(props.options[i].label)
      expect(CheckboxProps.value).toEqual(props.options[i].value)
      expect(CheckboxProps.name).toEqual(props.name)
      expect(CheckboxProps.required).toEqual(props.required)

      i += 1
    }
  })

  test('Value should match the checked Checkbox button', () => {
    // With no Checkbox selected
    const CheckboxComps = Checkboxes.getElements()
    let i = 0

    while (i < props.options.length) {
      const Checkbox = CheckboxComps[i]
      const CheckboxProps = Checkbox.props

      expect(CheckboxProps.checked).toBeFalsy()
      i += 1
    }

    // With the first Checkbox selected
    // (re-initialise the wrapper with changed props)
    const newProps = clone(props)
    newProps.value = ['one', 'three']

    const newWrapper = shallow(<CheckboxGroup {...newProps} />)
    const newCheckboxes = newWrapper.find(Checkbox)
    const newCheckboxComps = newCheckboxes.getElements()

    i = 0

    while (i < props.options.length) {
      const Checkbox = newCheckboxComps[i]
      const CheckboxProps = Checkbox.props

      if (i === 0 || i === 2) {
        expect(CheckboxProps.checked).toBeTruthy()
      } else {
        expect(CheckboxProps.checked).toBeFalsy()
      }

      i += 1
    }
  })

  test('onChange should be called and with correct values', () => {
    const newProps = clone(props)
    newProps.value = ['three']
    newProps.onChange = jest.fn()
    const newWrapper = shallow(<CheckboxGroup {...newProps} />)
    const checkboxTwo = newWrapper.find('Checkbox').at(1)
    checkboxTwo.simulate('change', {
      target: {
        value: 'two',
        checked: true,
      },
    })

    expect(newProps.onChange).toHaveBeenCalledWith(['three', 'two'])

    // Values in props should be the same after selection (no direct mutation)
    expect(newProps.value).toEqual(['three'])
  })
})
