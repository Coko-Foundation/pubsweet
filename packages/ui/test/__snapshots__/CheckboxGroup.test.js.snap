// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox Group is rendered correctly 1`] = `
<div
  className="ggGKkU"
>
  <label
    checked={false}
    className="fcsPpP"
  >
    <input
      checked={false}
      className="jGzFqE"
      onChange={[Function]}
      type="checkbox"
      value="one"
    />
    <span
      checked={false}
      className="eDCunu"
    >
      One
    </span>
  </label>
  <label
    checked={false}
    className="fcsPpP"
  >
    <input
      checked={false}
      className="jGzFqE"
      onChange={[Function]}
      type="checkbox"
      value="two"
    />
    <span
      checked={false}
      className="eDCunu"
    >
      Two
    </span>
  </label>
  <label
    checked={false}
    className="fcsPpP"
  >
    <input
      checked={false}
      className="jGzFqE"
      onChange={[Function]}
      type="checkbox"
      value="three"
    />
    <span
      checked={false}
      className="eDCunu"
    >
      Three
    </span>
  </label>
</div>
`;
